src/DeterministicNoiseSimulator.cpp

```diff
@@ -106,169 +122,214 @@ std::map<std::string, double> DeterministicNoiseSimulator::DeterministicSimulate
+            }
+        }
 
-    //if (line[density_op.p->v] > 0) {
-    if (op->actsOn(density_op.p->v)) {
-        //todo when A and D noise is active and prob is 0, I lose cached variables
         for (auto const& type: gateNoiseTypes) {
             switch (type) {
                 case 'A':
-                    ApplyAmplitudeDampingToNode(new_edges);
+                    ApplyAmplitudeDampingToNode(new_edges, (used_qubits.size() == 1) ? ampDampingProbSingleQubit : ampDampingProbMultiQubit);
                     break;
                 case 'P':
-                    ApplyPhaseFlipToNode(new_edges);
+                    ApplyPhaseFlipToNode(new_edges, (used_qubits.size() == 1) ? noiseProbSingleQubit : noiseProbMultiQubit);
                     break;
-                    //case 'B':
-                    //    applyBitFlipToNode(new_edges);
-                    //    break;
                 case 'D':
-                    ApplyDepolaritationToNode(new_edges);
+                    ApplyDepolarisationToNode(new_edges, (used_qubits.size() == 1) ? noiseProbSingleQubit : noiseProbMultiQubit);
                     break;
                 default:
                     throw std::runtime_error(std::string("Unknown gate noise type '") + type + "'");
             }
         }
+
+        for (auto& new_edge: new_edges) {
+            if (new_edge.w.approximatelyZero()) {
+                if (new_edge.w != dd::Complex::zero) {
+                    dd->cn.returnToCache(new_edge.w);
+                    new_edge.w = dd::Complex::zero;
+                }
+            } else {
+                dd::Complex c = dd->cn.lookup(new_edge.w);
+                dd->cn.returnToCache(new_edge.w);
+                new_edge.w = c;
+            }
+        }
     }
 
-    qc::MatrixDD tmp = dd->makeDDNode(density_op.p->v, new_edges, true);
+    e = dd->makeDDNode(originalCopy.p->v, new_edges, false, firstPathEdge);
+
+    // Adding the noise operation to the cache, note that e.w is from the complex number table
+    //    dd->densityNoise.insert(originalCopy, e, used_qubits);
 
-    // Multiplying the old edge weight with the new one
-    if (!tmp.w.approximatelyZero()) {
-        CN::mul(tmp.w, tmp.w, density_op.w);
+    // Multiplying the old edge weight with the new one and looking up in the complex numbers table
+    if (!e.w.approximatelyZero()) {
+        if (e.w.approximatelyOne()) {
+            e.w = originalEdge.w;
+        } else {
+            auto tmpComplexValue = dd->cn.getCached();
+            CN::mul(tmpComplexValue, e.w, originalEdge.w);
+            e.w = dd->cn.lookup(tmpComplexValue);
+            dd->cn.returnToCache(tmpComplexValue);
+        }
     }
-    return tmp;
+    return e;
 }
 
-void DeterministicNoiseSimulator::ApplyPhaseFlipToNode(std::array<qc::MatrixDD, 4>& e) {
-    double      probability  = noiseProbability;
+void DeterministicNoiseSimulator::ApplyPhaseFlipToNode(std::array<dEdge, 4>& e, double probability) {
```

Comment:
as already noted in the header, the below functions should reside in the DD Package as they only manipulate decision diagrams.

Reply:
The entire noise functionality is encapsulated into the QFR.